pymc_marketing/model_builder.py

```diff
@@ -277,42 +299,59 @@ def set_idata_attrs(
 
         Raises
         ------
+        ValueError
+            If the attrs are missing for a property initialization of the class
         RuntimeError
             If no InferenceData object is provided.
 
         Returns
         -------
-        None
+        InferenceData
+            The InferenceData instance with the attrs set
 
         Examples
         --------
-        >>> model = MyModel(ModelBuilder)
-        >>> idata = az.InferenceData(your_dataset)
-        >>> model.set_idata_attrs(idata=idata)
+        Set the attrs for an InferenceData object manually.
+
+        .. code-block:: python
+
+            idata: az.InferenceData = ...
+            model.set_idata_attrs(idata=idata)
+
         """
         if idata is None:
             idata = self.idata
         if idata is None:
             raise RuntimeError("No idata provided to set attrs on.")
 
-        def default(x):
-            if isinstance(x, Prior):
-                return x.to_json()
-            elif isinstance(x, HSGPKwargs):
-                return x.model_dump(mode="json")
-            return x.__dict__
+        attrs = self.create_idata_attrs()
+        attrs_keys = set(attrs.keys())
+        required_keys = {
+            "id",
+            "model_type",
+            "version",
+            "sampler_config",
+            "model_config",
+        }
+        if missing_keys := required_keys - attrs_keys:
+            msg = (
+                f"Missing required keys in attrs: {missing_keys}. "
+                "Call super().create_idata_attrs()."
+            )
+            raise ValueError(msg)
 
-        idata.attrs["id"] = self.id
-        idata.attrs["model_type"] = self._model_type
-        idata.attrs["version"] = self.version
-        idata.attrs["sampler_config"] = json.dumps(self.sampler_config)
-        idata.attrs["model_config"] = json.dumps(
-            self._serializable_model_config,
-            default=default,
-        )
-        # Only classes with non-dataset parameters will implement save_input_params
-        if hasattr(self, "_save_input_params"):
-            self._save_input_params(idata)
+        init_parameters: set[str] = set(signature(self.__init__).parameters.keys())  # type: ignore
+        # Remove since this will be stored in the fit_data group of InferenceData
+        init_parameters -= {"data"}
+
+        if missing_keys := init_parameters - attrs_keys:
+            msg = (
+                f"__init__ has parameters that are not in the attrs: {missing_keys}. "
+                "The save and load functionality will not work correctly."
+            )
+            raise ValueError(msg)
```

Comment:
Think we should do this step at initialization. WDYT? 
Then we'd know immediately if there is an issue with saving the model. Don't think this is too heavy of a calculation

Reply:
This actually has a bit of difficulty which the creation of the `id` which depends on the model_config attribute